numba/targets/arraymath.py

```diff
@@ -2663,3 +2663,27 @@ def impl(a, dtype=None):
                 return np.array(a, ty)
 
     return impl
+
+@overload(np.extract)
+def np_extract(condition, arr):
+
+    def np_extract_impl(condition, arr):
+        cond = np.asarray(condition).flatten()
+        a = np.asarray(arr)
+
+        if a.size == 0:
+            raise ValueError('Cannot extract from an empty array')
+
+        # the following looks odd but replicates NumPy...
+        if bool(cond[-1]) and cond.size > a.size:
```

Comment:
I'm not sure this check is strong enough, think it needs to walk the `condition` var from index `a.size` to `cond.size` looking for anything truthy. Consider the following...
```python
from numba import njit
import numpy as np

def foo(cond, arr):
    return np.extract(cond, arr)

arr = np.arange(4)
# cond[-1] is `False`, but cond[-2] is OOB
cond = np.array([True, False, True, False, False, True, False])

try:
    foo(cond, arr)
except Exception as e:
    print(e)
else:
    print("no exception raised")

try:
    njit(foo)(cond, arr)
except Exception as e:
    print(e)
else:
    print("no exception raised")
```
gives:
```
index 5 is out of bounds for size 4
no exception raised
```

Reply:
Oh, sorry - I will add more tests and fix it up.

Reply:
The last commit should hopefully cover it